### ISSUE TITLE
Do Not Contact is broken in Mautic 3.1

### ISSUE THREAD

**Issue by StoppingBuck** [mautic/mautic#9175]
## Bug Description
Do Not Contact no longer works as expected in 3.1, as there are multiple ways in which trying to set Do Not Contact does so with reason=0 (the corresponding row in the `lead_donotcontact` table in the DB gets a `0` in the `reason` column) thus actually rendering the contact 'Contactable' despite attempting to do the opposite.

I've found three ways in which this can happen:

- ~~Way 1: Calling the API endpoint for adding contact DNC to channel - `POST /contacts/<id>/dnc/email/add` - without a body (or with a body consisting of `{ "reason": 0 }`) causes the contact to get their DNC for the channel set to "Contactable", thus erroneously returning a response of 200 while actually not adding the contact to DNC, but merely adding a dummy row with reason=0 to the `lead_donotcontact` table. This can even be spammed to create an enormous amount of dummy rows (#9076) 
Furthermore, this runs counter to https://developer.mautic.org/#add-do-not-contact which for `reason` states that "Default is Manual", which is not currently the case. The documentation is right, the code is wrong: Default reason should be `3`, not `0`.
I'd argue that in addition to changing the default reason from `0` ==> `3`, trying to add a DNC with `{ "reason": 0 }` should cause an error, as it is misleading to return a 200 OK response from an endpoint that explicitly exists to **add** DNC if that does not actually happen.~~ (moved to a separate bug report: #9076 )

- Way 2: Setting up Monitored Inbox for automatic unsubscribe-emails also results in them being set reason=0 with the 1 accidentally being set in the comments column instead of the reason column.

- Way 3: Going to the contacts list view - `s/contacts` - and checking one or more contacts and then selecting 'Set Do Not Contact' and inputting whatever reason you want also accidentally gives them reason=0 and puts the 3 that's supposed to indicate 'Manual' as part of the `comments` column. 

This is likely due to an errant comma somewhere in the code which pushes the reason integer to the comment column in the DB instead of the reason column.

---
| Q   | A
| --- | ---
| Mautic version | 3.1
| PHP version | 7.3

**Comment by StoppingBuck** [comment]
The PR #8434 might potentially solve this. Needs to be tested.

**Comment by npracht** [comment]
Can you test it ?

**Comment by StoppingBuck** [comment]
@npracht I just tested the PR with Mautibox and unfortunately that PR does not seem to fix this big - calling `POST https://mautibox.com/8434/api/contacts/50/dnc/email/add` still results in the contact getting a DNC with reason=0 (contactable) instead of a DNC with reason=3 (manual). 

Can you please remove the `ready-to-test` tag?

**Comment by StoppingBuck** [comment]
This one seems to be bigger than just the API, so I'm updating the description. I've identified two other situations where the same can happen:

- Setting up Monitored Inbox for automatic unsubscribe-emails also results in them being set reason=0 with the 1 accidentally being set in the comments column instead of the reason column.

- Going to the contacts list view - `s/contacts` - and checking one or more contacts and then selecting 'Set Do Not Contact' and inputting whatever reason you want also accidentally gives them reason=0 and puts the 3 that's supposed to indicate 'Manual' as part of the `comments` column. 

I am 95% sure that this is due to an errant comma somewhere in the code that pushes the reason number into the comments column by accident. If nobody takes this before then, I might try tackling it on the sprint.

**Comment by StoppingBuck** [comment]
Here's a sample query from `lead_donotcontact` table to show what I mean. The ones with comment '1' were from Monitored Inbox, the ones with comment '3' were manually set DoNotContact from the Mautic interface:

```
+--------+---------+---------------------+--------+---------+------------+----------+
| id     | lead_id | date_added          | reason | channel | channel_id | comments |
+--------+---------+---------------------+--------+---------+------------+----------+
| 157416 |   27873 | 2020-09-08 17:58:40 |      0 | email   |       NULL | 3        |
| 160300 |  283215 | 2020-09-08 17:59:35 |      0 | email   |       NULL | 3        |
| 166690 |  372884 | 2020-09-08 17:59:07 |      0 | email   |       NULL | 3        |
| 190090 |   46683 | 2020-09-08 18:00:06 |      0 | email   |       NULL | 3        |
| 193960 |  370910 | 2020-09-08 17:50:35 |      0 | email   |       NULL | 3        |
| 193962 |  139015 | 2020-09-08 17:51:07 |      0 | email   |       NULL | 3        |
| 193965 |  103690 | 2020-09-08 17:51:39 |      0 | email   |       NULL | 3        |
| 193966 |   97473 | 2020-09-08 17:52:12 |      0 | email   |       NULL | 3        |
| 193967 |  103690 | 2020-09-08 17:52:39 |      0 | email   |       NULL | 3        |
| 193969 |  454479 | 2020-09-08 17:53:12 |      0 | email   |       NULL | 3        |
| 193971 |  114004 | 2020-09-08 17:53:43 |      0 | email   |       NULL | 3        |
| 193973 |  332697 | 2020-09-08 17:54:23 |      0 | email   |       NULL | 3        |
| 193975 |  122806 | 2020-09-08 17:54:56 |      0 | email   |       NULL | 3        |
| 193978 |  175176 | 2020-09-08 17:55:32 |      0 | email   |       NULL | 3        |
| 193979 |  350690 | 2020-09-08 17:56:02 |      0 | email   |       NULL | 3        |
| 193981 |   86518 | 2020-09-08 17:56:35 |      0 | email   |       NULL | 3        |
| 193984 |  564864 | 2020-09-08 17:57:08 |      0 | email   |       NULL | 3        |
| 193986 |  133055 | 2020-09-08 17:57:41 |      0 | email   |       NULL | 3        |
| 193988 |  135382 | 2020-09-08 17:58:10 |      0 | email   |       NULL | 3        |
| 193997 |  586222 | 2020-09-08 18:00:39 |      0 | email   |       NULL | 3        |
| 193998 |  373888 | 2020-09-08 18:01:15 |      0 | email   |       NULL | 3        |
| 194266 |  120008 | 2020-09-09 08:21:06 |      0 | email   |        870 | 1        |
| 194267 |   65857 | 2020-09-09 08:21:07 |      0 | email   |        870 | 1        |
| 194268 |   75218 | 2020-09-09 08:21:08 |      0 | email   |        870 | 1        |
| 194269 |  203868 | 2020-09-09 08:21:09 |      0 | email   |        870 | 1        |
| 194270 |  558700 | 2020-09-09 08:21:09 |      0 | email   |        870 | 1        |
| 194271 |   74755 | 2020-09-09 08:21:10 |      0 | email   |        870 | 1        |
| 194272 |   88958 | 2020-09-09 08:21:10 |      0 | email   |        870 | 1        |
| 194273 |  531874 | 2020-09-09 08:21:11 |      0 | email   |        870 | 1        |
| 194274 |  120517 | 2020-09-09 08:21:11 |      0 | email   |        870 | 1        |
| 194275 |  517438 | 2020-09-09 08:21:12 |      0 | email   |        870 | 1        |
| 194278 |  564864 | 2020-09-09 08:34:35 |      0 | email   |       NULL | 3        |
| 194280 |  373888 | 2020-09-09 08:36:46 |      0 | email   |       NULL | 3        |
+--------+---------+---------------------+--------+---------+------------+----------+
```

**Comment by StoppingBuck** [comment]
Alright, some sleuthing later and here's some findings:

- [This line](https://github.com/mautic/mautic/blob/8cb3c509a09224fe9c71abe84faebdb69c11c934/app/bundles/EmailBundle/MonitoredEmail/Processor/FeedbackLoop.php#L96) has reason and comments the wrong way around compared to what `addDncForContact()` expects. The change was part of [this commit](https://github.com/mautic/mautic/commit/210bd3b0356340499a62d62567bcb6347ccea765) by @p365labs 7 months ago.

- Likewise, [this line](https://github.com/mautic/mautic/blob/8cb3c509a09224fe9c71abe84faebdb69c11c934/app/bundles/SmsBundle/Helper/SmsHelper.php#L108) also seems to have comments and reason swapped compared to what is expected.

- [This line](https://github.com/mautic/mautic/blob/8cb3c509a09224fe9c71abe84faebdb69c11c934/app/bundles/EmailBundle/MonitoredEmail/Processor/Unsubscribe.php#L119) also seems swapped

- I can't really decipher what's going on on [this line](https://github.com/mautic/mautic/blob/8cb3c509a09224fe9c71abe84faebdb69c11c934/app/bundles/LeadBundle/EventListener/CampaignActionDNCSubscriber.php#L90). It seems like the reason is being sent as comments, but since the literal for MANUAL reason is passed as reason, it shouldn't really matter. Still weird.

I can't say for sure whether the above lines are the cause of all these problems, but it seems that fixing them might at the very least fix the Monitored Inbox part of this bug.

**Comment by StoppingBuck** [comment]
I didn't notice that my search had multiple pages :facepalm: So here are some more findings:

- [This line](https://github.com/mautic/mautic/blob/8cb3c509a09224fe9c71abe84faebdb69c11c934/app/bundles/EmailBundle/MonitoredEmail/Processor/Bounce.php#L144) includes comments, but not reason, likely causing the comments to be used as reason (if my PHP-fu is correct)

- As for the API part, I think it's being controlled [here](https://github.com/mautic/mautic/blob/1d57fb7d872c5bf35158af915d468ff0731f470c/app/bundles/LeadBundle/Controller/Api/LeadApiController.php#L418) so any check for reason=0 should probably be added around that part.

- [This line](https://github.com/mautic/mautic/blob/8cb3c509a09224fe9c71abe84faebdb69c11c934/app/bundles/LeadBundle/Controller/AjaxController.php#L357) also has reason and comment swapped, and may be the reason for the weirdness when setting DNC from the Mautic interface

- [This line](https://github.com/mautic/mautic/blob/d2018fb96beeec25eae200caa286444e12608ba8/app/bundles/EmailBundle/Model/EmailModel.php#L1763) also has reason and comment swapped

- I'm not sure about [this line](https://github.com/mautic/mautic/blob/5ca3b842fcc57dc5c73b5b06a37fcb1dc19ef551/app/bundles/LeadBundle/Controller/LeadController.php#L1643) but it looks swapped too.

- [This line](https://github.com/mautic/mautic/blob/25c203b54527d3636c5cad917d178e32a02bd4d6/plugins/MauticCrmBundle/Integration/SalesforceIntegration.php#L2717) is swapped as well.

My guess is that for all or most of the above the reason and comments parameters need to be swapped.

**Comment by StoppingBuck** [comment]
I will try making a PR later today.

**Comment by RCheesley** [comment]
Thanks for all the sleuthing @StoppingBuck 🕵️  ! 

Seems we have some rather confused code knocking about - I bet there are some other issues that have the same root cause just looking at what you have found above.

I think given the scope of this we will end up closing https://github.com/mautic/mautic/issues/9076 in favour of this one.

Looking forward to the PR - thanks for taking this on!

**Comment by npracht** [comment]
Hi there, can you please test https://github.com/mautic/mautic/pull/9186 ?

**Comment by StoppingBuck** [comment]
@npracht It is a step in the right direction, :+1:  but lacks a few fixes to completely eliminate the problem. I've left a comment on the PR itself.

**Comment by StoppingBuck** [comment]
#9186 has been expanded and now fixes this bug. Once that PR is merged, this report can be closed.

**Comment by dennisameling** [comment]
Hi! To keep things central regarding issues around Do Not Contact (DNC), I created a JIRA issue to track all issues related to DNC in a single place. That way, we can more efficiently fix the several bugs that were reported. Please check it out: https://mautic.atlassian.net/browse/TPROD-72

**Comment by npracht** [comment]
Fixed in 3.1.2
